File: testdata/fixtures/js_clones/productService.ts
import { db } from './db';
import { NotFoundError } from './errors';

interface Product {
  id: string;
  title: string;
  price: number;
}

export async function getProduct(id: string): Promise<Product> {
  const product = await db.query('SELECT * FROM products WHERE id = $1', [id]);
  if (!product) {
    throw new NotFoundError(`Product not found: ${id}`);
  }
  if (!product.active) {
    throw new Error('Product is inactive');
  }
  return {
    id: product.id,
    title: product.title,
    price: product.price,
  };
}

export async function updateProduct(id: string, data: Partial<Product>): Promise<Product> {
  const product = await db.query('SELECT * FROM products WHERE id = $1', [id]);
  if (!product) {
    throw new NotFoundError(`Product not found: ${id}`);
  }
  if (!product.active) {
    throw new Error('Product is inactive');
  }
  const updated = await db.query('UPDATE products SET title = $1, price = $2 WHERE id = $3 RETURNING *', [
    data.title || product.title,
    data.price || product.price,
    id,
  ]);
  return {
    id: updated.id,
    title: updated.title,
    price: updated.price,
  };
}

export async function deleteProduct(id: string): Promise<void> {
  const product = await db.query('SELECT * FROM products WHERE id = $1', [id]);
  if (!product) {
    throw new NotFoundError(`Product not found: ${id}`);
  }
  if (!product.active) {
    throw new Error('Product is inactive');
  }
  await db.query('DELETE FROM products WHERE id = $1', [id]);
}

File: testdata/fixtures/rb_clones/product_service.rb
class ProductService
  def initialize(repo)
    @repo = repo
  end

  def find_product(id)
    product = @repo.find(id)
    raise NotFoundError, "Product not found: #{id}" unless product
    raise InactiveError, "Product is inactive" unless product.active?
    product
  end

  def update_product(id, attrs)
    product = @repo.find(id)
    raise NotFoundError, "Product not found: #{id}" unless product
    raise InactiveError, "Product is inactive" unless product.active?
    product.update(attrs)
    @repo.save(product)
    product
  end

  def delete_product(id)
    product = @repo.find(id)
    raise NotFoundError, "Product not found: #{id}" unless product
    raise InactiveError, "Product is inactive" unless product.active?
    @repo.delete(product)
  end
end
